File: macos/Tests/ElevateCoreTests/Support/StubHTTPClient.swift
import Foundation
import ElevateCore

/// Routes requests by HTTP method plus a substring of the URL. Records every request.
actor StubHTTPClient: HTTPClient {
    struct Route { let method: String; let urlContains: String; let respond: @Sendable (HTTPRequest) async throws -> HTTPResponse }
    private var routes: [Route] = []
    private(set) var requests: [HTTPRequest] = []

    func on(_ method: String, _ urlContains: String, status: Int = 200, headers: [String: String] = [:], body: Data = Data()) {
        routes.append(Route(method: method, urlContains: urlContains) { _ in HTTPResponse(status: status, headers: headers, body: body) })
    }

    func on(_ method: String, _ urlContains: String, respond: @escaping @Sendable (HTTPRequest) -> HTTPResponse) {
        routes.append(Route(method: method, urlContains: urlContains, respond: respond))
    }

    /// A route that never answers: the request ends only when the caller's own deadline cancels
    /// it. Used to prove a bounded fetch is bounded.
    func hang(_ method: String, _ urlContains: String) {
        routes.append(Route(method: method, urlContains: urlContains) { _ in
            try await Task.sleep(for: .seconds(600))
            return HTTPResponse(status: 599)
        })
    }

    func send(_ request: HTTPRequest) async throws -> HTTPResponse {
        requests.append(request)
        guard let route = routes.last(where: { $0.method == request.method && request.url.absoluteString.contains($0.urlContains) }) else {
            return HTTPResponse(status: 599, body: Data("no stub for \(request.method) \(request.url)".utf8))
        }
        return try await route.respond(request)
    }

    func requests(matching substring: String) -> [HTTPRequest] {
        requests.filter { $0.url.absoluteString.contains(substring) }
    }
}

File: audit/tests/Elevate.Audit.Tests/Support/StubHttpClient.cs
using System.Text;
using Elevate.Core.Networking;

namespace Elevate.Audit.Tests.Support;

/// <summary>Routes requests by HTTP method plus a URL substring; the last matching registration wins. Records every request.</summary>
public sealed class StubHttpClient : IHttpClient
{
    private sealed record Route(string Method, string UrlContains, Func<HttpRequestData, Task<HttpResponseData>> Respond);

    private readonly List<Route> _routes = [];
    private readonly List<HttpRequestData> _requests = [];
    private readonly Lock _gate = new();

    public IReadOnlyList<HttpRequestData> Requests
    {
        get { lock (_gate) { return [.. _requests]; } }
    }

    public IReadOnlyList<HttpRequestData> RequestsMatching(string urlContains) =>
        Requests.Where(r => Unescaped(r.Url).Contains(urlContains, StringComparison.Ordinal)).ToList();

    public void On(string method, string urlContains, string body, int status = 200) =>
        On(method, urlContains, _ => new HttpResponseData(status, new Dictionary<string, string> { ["Content-Type"] = "application/json" }, Encoding.UTF8.GetBytes(body)));

    public void On(string method, string urlContains, Func<HttpRequestData, HttpResponseData> respond) =>
        On(method, urlContains, request => Task.FromResult(respond(request)));

    /// <summary>For tests that need a response to complete out of order relative to sibling requests (e.g. bounded-concurrency waves), such as one that awaits a delay before responding.</summary>
    public void On(string method, string urlContains, Func<HttpRequestData, Task<HttpResponseData>> respond)
    {
        lock (_gate) { _routes.Add(new Route(method, urlContains, respond)); }
    }

    public Task<HttpResponseData> SendAsync(HttpRequestData request, CancellationToken ct)
    {
        Route? route;
        lock (_gate)
        {
            _requests.Add(request);
            route = _routes.LastOrDefault(r =>
                string.Equals(r.Method, request.Method, StringComparison.OrdinalIgnoreCase)
                && Unescaped(request.Url).Contains(r.UrlContains, StringComparison.Ordinal));
        }

        return route is null
            ? Task.FromResult(new HttpResponseData(599, new Dictionary<string, string>(), Encoding.UTF8.GetBytes($"no stub for {request.Method} {request.Url}")))
            : route.Respond(request);
    }

    /// <summary>
    /// Matches against the decoded URL: OData filters such as <c>groupId eq 'g1'</c> are registered with
    /// literal spaces and quotes, while <see cref="Uri.AbsoluteUri"/> percent-encodes them.
    /// </summary>
    private static string Unescaped(Uri url) => Uri.UnescapeDataString(url.AbsoluteUri);
}
